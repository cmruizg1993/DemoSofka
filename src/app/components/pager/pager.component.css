.pages>ul{
    display: inline;
}
.pages>ul>li{
    display: inline;
    margin-left: 1rem;
    cursor: pointer;
}
.pages>ul>li>span{
    font-size: 1rem;
    color: var(--blue-600);
}
.pages>ul>li>span:hover{
    color: var(--blue-700);
    text-decoration: underline;
}
/*
.pages>ul>li>span.active{
    font-weight: 800;
}
*/