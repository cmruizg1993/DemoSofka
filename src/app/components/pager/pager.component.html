<div class="row p-2">
    <div class="col">
        <p class="fs-2 fw-800 fs-2 fw-800">Mostrando {{pagerProperties.pageSize}} de {{dataLength}}&emsp; Resultados</p>
    </div>
    <div class="col d-flex justify-content-end">
        <div class="pages">
            <ul>
                <li  *ngFor="let page of pages" (click)="updatePage(page)">
                    <span [class.active]="page.isActive">{{page.number}}</span>
                </li>
            </ul>
            <p class="text-primary fs-2 fw-800">Página actual: {{pagerProperties.page}}</p>
        </div>
    </div>
    <div class="col d-flex justify-content-end">
        <app-select 
        [data]="[{label: '5', value: 5, isSelected: true},{label: '10', value: 10},{label: '20', value: 20}]"
        label="Items por pág."
        (modelUpdate)="updatePageSize($event)">
        </app-select>
    </div>
</div>

