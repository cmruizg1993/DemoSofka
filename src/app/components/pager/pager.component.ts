import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'app-pager',
  templateUrl: './pager.component.html',
  styleUrl: './pager.component.css'
})
export class PagerComponent {
  @Output() pagerUpdate =  new EventEmitter<PagerProperties>();
  @Input() dataLength = 0;
  
  pagerProperties = new PagerProperties();
  
  pages: Page[] = [];
  ngOnInit(): void {
    //Called after the constructor, initializing input properties, and the first call to ngOnChanges.
    //Add 'implements OnInit' to the class.
    
  }
  ngAfterContentChecked(): void {
    //Called after every check of the component's or directive's content.
    //Add 'implements AfterContentChecked' to the class.
    if(this.dataLength > 0){
      const numberOfPages = Math.ceil(this.dataLength/this.pagerProperties.pageSize);
      this.pages = Array(numberOfPages).fill(null).map((x,i)=>{
        const page = new Page();
        page.number = i+1;
        page.isActive = i == 0 ;
        return page;
      });
      //console.log(this.pages)
    }
  }
  updatePage(page: Page){
    this.pagerProperties.page = page.number;
    this.updatePager();
  }
  updatePageSize(event: any){
    this.pagerProperties.page = 1;
    this.pagerProperties.pageSize = event.value;
    this.updatePager();
  }
  updatePager(){
    this.pagerUpdate.emit(this.pagerProperties);
  }
  
}
class PagerProperties{
  page: number = 1;
  pageSize = 5;
}
class Page{
  number: number = 0;
  isActive: boolean = false;
}
