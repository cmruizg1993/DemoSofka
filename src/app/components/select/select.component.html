<label *ngIf="label">
    {{label}}
</label>
<select [(ngModel)]="model">
    <option [ngValue]="item" *ngFor="let item of data">{{item.label}}</option>
</select>
