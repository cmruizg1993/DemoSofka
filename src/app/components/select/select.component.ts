import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'app-select',
  templateUrl: './select.component.html',
  styleUrl: './select.component.css'
})
export class SelectComponent {
  @Input() data: any[] = [];
  @Input() label = '';
  default: any;
  @Output() modelUpdate =  new EventEmitter<any>();
  private privateModel: any;

  ngOnInit(): void {
    //Called after the constructor, initializing input properties, and the first call to ngOnChanges.
    //Add 'implements OnInit' to the class.
    this.default = this.data.find( d => d.isSelected );
    if(this.default)
        this.model = this.default;
  }
  set model(newModel: any){
    this.privateModel = newModel;
    this.emitUpdate();
  }
  get model(){
    return this.privateModel;
  }
  
  emitUpdate(){
    //console.log(this.model)
    this.modelUpdate.emit(this.model);
  }
}
