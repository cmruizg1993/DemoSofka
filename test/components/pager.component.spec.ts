import { ComponentFixture, TestBed } from '@angular/core/testing';

import { RouterTestingModule } from '@angular/router/testing';
import { PagerComponent } from '../../src/app/components/pager/pager.component';
import { SelectComponent } from '../../src/app/components/select/select.component';
import { FormsModule } from '@angular/forms';



describe('PagerComponent', () => {
  let fixture: ComponentFixture<PagerComponent>;
  let compiled: HTMLElement;
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        FormsModule
      ],
      declarations: [
        PagerComponent,
        SelectComponent,        
      ],
    }).compileComponents();
    fixture = TestBed.createComponent(PagerComponent);
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('Se debe crear el componente', () => {
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });
  it(`Debe coincidir con el snapshot'`, () => {
    expect(compiled).toMatchSnapshot();
  });
})