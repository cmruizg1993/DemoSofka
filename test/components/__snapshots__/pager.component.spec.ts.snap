// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagerComponent Debe coincidir con el snapshot' 1`] = `
<div
  id="root1"
>
  <div
    class="row p-2"
  >
    <div
      class="col"
    >
      <p
        class="fs-2 fw-800 fs-2 fw-800"
      >
        Mostrando 5 de 0 Resultados
      </p>
    </div>
    <div
      class="col d-flex justify-content-end"
    >
      <div
        class="pages"
      >
        <ul>
          
        </ul>
        <p
          class="text-primary fs-2 fw-800"
        >
          Página actual: 1
        </p>
      </div>
    </div>
    <div
      class="col d-flex justify-content-end"
    >
      <app-select
        label="Items por pág."
      >
        <label>
           Items por pág.

        </label>
        
        <select
          class="ng-untouched ng-pristine ng-valid"
        >
          <option
            value="0: Object"
          >
            5
          </option>
          <option
            value="1: Object"
          >
            10
          </option>
          <option
            value="2: Object"
          >
            20
          </option>
          
        </select>
      </app-select>
    </div>
  </div>
</div>
`;
